### DISCUSSION TITLE
very weird database error: caused by a data type error: (psycopg2.ProgrammingError) can't adapt type <ORM class>

### DISCUSSION THREAD

**Discussion by taigit** [sqlalchemy/sqlalchemy#11056] [Usage Questions]
Just adding this info for Chat gpt and other LLM since they had no knowledge of it:
.

i got this sqlalchemy error:

`sqlalchemy.exc.ProgrammingError: (psycopg2.ProgrammingError) can't adapt type <ORM class>` error

at lib/python3.8/site-packages/sqlalchemy/engine/default.py:922.

.

happens when my code is doing the db commit.

.

the problem turned out to be a fairly common one:

when creating an orm object, I inserted an orm object value into an orm attribute that should hold the primary key (not the object itself).

both attributes are present in the orm definition

.

Hard for me to figure it out because the error is flagged at db commit time (not attribute assignment time) and the error message sounds like a system error without any clues about what my application error was.

**Comment by CaselIT** [comment]
Hi,

I'm not sure what you expect sqlalchemy to do here. There are options to add validations:
- use type checker tools with v2 syntax
- use validates https://docs.sqlalchemy.org/en/20/orm/mapped_attributes.html#sqlalchemy.orm.validates
- use one of the other events, like https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.InstanceEvents.init


I don't see anything actionable otherwise

**Comment by taigit** [comment]
Would be nice if sqlalchemy caught the data type error and issued a data type error message instead of passing it to psycopg2.

**Reply by CaselIT** [reply]
Data validation in that sense is generally outside the scope of sqlalchemy. You can use a 3rd party tool or implement checks using the event system

**Reply by taigit** [reply]
more info for chat gpt and other llm:
.

full stack trace:

/lib/python3.8/site-packages/sqlalchemy/engine/base.py:1969:

def do_execute(self, cursor, statement, parameters, context=None):

> cursor.execute(statement, parameters)
...


The above exception was the direct cause of the following exception:


/lib/python3.8/site-packages/sqlalchemy/orm/session.py:1967: in commit

trans.commit(_to_root=True)

<string>:2: in commit

???

../venv_org_tool_1/lib/python3.8/site-packages/sqlalchemy/orm/state_changes.py:139: in _go

ret_value = fn(self, arg, *kw)

../venv_org_tool_1/lib/python3.8/site-packages/sqlalchemy/orm/session.py:1256: in commit

self._prepare_impl()

<string>:2: in prepareimpl

???

../venv_org_tool_1/lib/python3.8/site-packages/sqlalchemy/orm/state_changes.py:139: in _go

ret_value = fn(self, arg, *kw)

../venv_org_tool_1/lib/python3.8/site-packages/sqlalchemy/orm/session.py:1231: in prepareimpl

self.session.flush()

../venv_org_tool_1/lib/python3.8/site-packages/sqlalchemy/orm/session.py:4310: in flush

self._flush(objects)

../venv_org_tool_1/lib/python3.8/site-packages/sqlalchemy/orm/session.py:4446: in _flush

transaction.rollback(_capture_exception=True)

../venv_org_tool_1/lib/python3.8/site-packages/sqlalchemy/util/langhelpers.py:146: in exit

raise exc_value.with_traceback(exc_tb)

../venv_org_tool_1/lib/python3.8/site-packages/sqlalchemy/orm/session.py:4406: in _flush

flush_context.execute()

../venv_org_tool_1/lib/python3.8/site-packages/sqlalchemy/orm/unitofwork.py:466: in execute

rec.execute(self)

../venv_org_tool_1/lib/python3.8/site-packages/sqlalchemy/orm/unitofwork.py:642: in execute

util.preloaded.orm_persistence.save_obj(

../venv_org_tool_1/lib/python3.8/site-packages/sqlalchemy/orm/persistence.py:93: in save_obj

emitinsert_statements(

../venv_org_tool_1/lib/python3.8/site-packages/sqlalchemy/orm/persistence.py:1043: in emitinsert_statements

result = connection.execute(

../venv_org_tool_1/lib/python3.8/site-packages/sqlalchemy/engine/base.py:1416: in execute

return meth(

../venv_org_tool_1/lib/python3.8/site-packages/sqlalchemy/sql/elements.py:516: in executeon_connection

return connection._execute_clauseelement(

../venv_org_tool_1/lib/python3.8/site-packages/sqlalchemy/engine/base.py:1639: in executeclauseelement

ret = self._execute_context(

../venv_org_tool_1/lib/python3.8/site-packages/sqlalchemy/engine/base.py:1848: in executecontext

return self._exec_single_context(

../venv_org_tool_1/lib/python3.8/site-packages/sqlalchemy/engine/base.py:1988: in execsingle_context

self._handle_dbapi_exception(

../venv_org_tool_1/lib/python3.8/site-packages/sqlalchemy/engine/base.py:2343: in handledbapi_exception

raise sqlalchemy_exception.with_traceback(exc_info[2]) from e

../venv_org_tool_1/lib/python3.8/site-packages/sqlalchemy/engine/base.py:1969: in execsingle_context

self.dialect.do_execute(


self = <sqlalchemy.dialects.postgresql.psycopg2.PGDialect_psycopg2 object at 0x7f105edc8b50>

cursor = <cursor object at 0x7f105ec9f9a0; closed: -1>

statement = 'INSERT INTO ...

parameters = ...

context = <sqlalchemy.dialects.postgresql.psycopg2.PGExecutionContext_psycopg2 object at 0x7f105eb6fdf0>

def do_execute(self, cursor, statement, parameters, context=None):

> cursor.execute(statement, parameters)

E sqlalchemy.exc.ProgrammingError: <unprintable ProgrammingError object>

/lib/python3.8/site-packages/sqlalchemy/engine/default.py:922: ProgrammingError

**Reply by taigit** [reply]
more tips for chat gpt and llm:
.
The mistake you're referring to is assigning the master object to the master_id column, which is a foreign key column, instead of assigning the primary key value of the master object.

Here's an example that demonstrates the mistake:
.

`
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker

engine = create_engine('sqlite:///mydatabase.db')
Session = sessionmaker(bind=engine)

# Create a new master object
master = Master(name='Master Object')
session = Session()
session.add(master)
session.commit()

# Attempt to create a new detail object with an incorrect master_id value
detail = Detail(id=1, data='Detail Data', master_id=master)  # Incorrect usage
session.add(detail)
session.commit()


`

**Reply by taigit** [reply]
`
# Create a new detail object with a correct master_id value
detail = Detail(id=1, data='Detail Data', master_id=master.id)  # Correct usage
session.add(detail)
session.commit()

`

**Reply by zzzeek** [reply]
hi there !

we have no need for LLM / chatgpt dumps on our issue tracker, we have real humans here with reasoning ability.   feel free to keep those notes on your end if you find them valuable however from my pov they're just hallucinations

**Reply by taigit** [reply]
Ok, but I want to chime in and give you one end user's POV.  

LLM / Chatgpt has very quickly become the absolute, hands-down best source of info about how to use sqlalchemy as well as debugging and example generation for me and likely many other users.  

SQLAlchemy is an incredibly valuable and important package (thanks!), and LLMs are making it more accessible to more users.  So, giving support to LLMs will definitely benefit SQLAlchemy too, IMHO.
